katcp/core.py

```diff
@@ -293,22 +305,26 @@ def __init__(self, mtype, name, arguments=None, mid=None):
                                    % (name,))
 
     def format_argument(self, arg):
-        """Format a Message argument to a string"""
-        if isinstance(arg, float):
-            return repr(arg)
-        elif isinstance(arg, bool):
-            return str(int(arg))
+        """Format a Message argument to a byte string"""
+        if isinstance(arg, bool):
+            return b'1' if arg else b'0'
+        elif isinstance(arg, int):
+            return b'%d' % arg
+        elif isinstance(arg, float):
+            return repr(arg).encode('ascii')
+        elif future.utils.isbytes(arg):
+            return arg
         else:
             try:
-                return str(arg)
+                if not future.utils.istext(arg):
+                    arg = str(arg)
+                return arg.encode('utf-8')
```

Comment:
> Disagree. For a unicode input, if not future.utils.istext(arg) will return false (it is text), in Python 2 (and 3), so we won't call str(arg).

You're right.

I guess what's bugging me is that on Python 2, you're effectively saying: "if it's not unicode, coerce it to bytes`, which means that after the `if` statement you might have either bytes or unicode.

> It would be nice to remove the try-except, but in the test cases I have this weird string which isn't encodable

Yuck, I hadn't realised Python would even let you create a Unicode string with surrogates in it.